src/validationinterface.cpp

```diff
@@ -16,36 +16,59 @@
 
-#include <boost/signals2/signal.hpp>
-
-struct ValidationInterfaceConnections {
-    boost::signals2::scoped_connection UpdatedBlockTip;
-    boost::signals2::scoped_connection TransactionAddedToMempool;
-    boost::signals2::scoped_connection BlockConnected;
-    boost::signals2::scoped_connection BlockDisconnected;
-    boost::signals2::scoped_connection TransactionRemovedFromMempool;
-    boost::signals2::scoped_connection ChainStateFlushed;
-    boost::signals2::scoped_connection BlockChecked;
-    boost::signals2::scoped_connection NewPoWValidBlock;
-};
-
 struct MainSignalsInstance {
-    boost::signals2::signal<void (const CBlockIndex *, const CBlockIndex *, bool fInitialDownload)> UpdatedBlockTip;
-    boost::signals2::signal<void (const CTransactionRef &)> TransactionAddedToMempool;
-    boost::signals2::signal<void (const std::shared_ptr<const CBlock> &, const CBlockIndex *pindex)> BlockConnected;
-    boost::signals2::signal<void (const std::shared_ptr<const CBlock>&, const CBlockIndex* pindex)> BlockDisconnected;
-    boost::signals2::signal<void (const CTransactionRef &)> TransactionRemovedFromMempool;
-    boost::signals2::signal<void (const CBlockLocator &)> ChainStateFlushed;
-    boost::signals2::signal<void (const CBlock&, const BlockValidationState&)> BlockChecked;
-    boost::signals2::signal<void (const CBlockIndex *, const std::shared_ptr<const CBlock>&)> NewPoWValidBlock;
-
     // We are not allowed to assume the scheduler only runs in one thread,
     // but must ensure all callbacks happen in-order, so we end up creating
     // our own queue here :(
     SingleThreadedSchedulerClient m_schedulerClient;
-    std::unordered_map<CValidationInterface*, ValidationInterfaceConnections> m_connMainSignals;
+    Mutex m_mutex;
+    struct Entry { std::shared_ptr<CValidationInterface> callbacks; int count = 1; };
+    std::list<Entry> m_list GUARDED_BY(m_mutex);
+    std::unordered_map<CValidationInterface*, std::list<Entry>::iterator> m_map GUARDED_BY(m_mutex);
 
     explicit MainSignalsInstance(CScheduler *pscheduler) : m_schedulerClient(pscheduler) {}
+
+    void Add(std::shared_ptr<CValidationInterface> callbacks)
+    {
+        LOCK(m_mutex);
+        auto inserted = m_map.emplace(callbacks.get(), m_list.end());
+        if (inserted.second) {
+            m_list.emplace_back();
+            inserted.first->second = std::prev(m_list.end());
+        }
+        inserted.first->second->callbacks = std::move(callbacks);
+    }
+
+    void Remove(CValidationInterface* key)
+    {
+        LOCK(m_mutex);
+        auto it = m_map.find(key);
+        if (it != m_map.end()) {
+            if (--it->second->count == 0) m_list.erase(it->second);
+            m_map.erase(it);
+        }
+    }
+
+    void Clear()
+    {
+        LOCK(m_mutex);
+        for (auto it = m_list.begin(); it != m_list.end();) {
+            if (--it->count == 0) it = m_list.erase(it); else ++it;
+        }
+        m_map.clear();
+    }
```

Comment:
It follows from the code of the `Clear()` function that `m_list` could be non-empty when the function returns.

Is it consistent state if `m_list.empty() == false` and `m_map.empty() == true` ?

Reply:
re: https://github.com/bitcoin/bitcoin/pull/18524#discussion_r403687245

> It follows from the code of the `Clear()` function that `m_list` could be non-empty when the function returns.
> 
> Is it consistent state if `m_list.empty() == false` and `m_map.empty() == true` ?

Yes, this happens when the reference count of one or more callbacks in the list is nonzero, which means they are currently being called, and will be removed by the iterate function when they are done being called, clearing the list.